llvm/include/llvm/Transforms/Instrumentation/CFGMST.h

```diff
@@ -92,6 +94,39 @@ template <class Edge, class BBInfo> class CFGMST {
     return It->second.get();
   }
 
+  void handleCoroSuspendEdge(Edge *E) {
+    // We must not add instrumentation to the BB representing the
+    // "suspend" path, else CoroSplit won't be able to lower
+    // llvm.coro.suspend to a tail call. We do want profiling info for
+    // the other branches (resume/destroy). So we do 2 things:
+    // 1. we prefer instrumenting those other edges by setting the weight
+    //    of the "suspend" edge to max, and
+    // 2. we mark the edge as "Removed" to guarantee it is not considered
+    //    for instrumentation. That could technically happen:
+    //    (from test/Transforms/Coroutines/coro-split-musttail.ll)
+    //
+    // %suspend = call i8 @llvm.coro.suspend(token %save, i1 false)
```

Comment:
how is suspend lowered? will a tail call be inserted in the dest BB?

Reply:
IIUC, `llvm.coro.suspend` is really a marker that says "return. Upon resumption, here's where you go if there's more to execute, and here's where you go if we're done". So it's not really a call, and the switch isn't really a switch, and the "default" path of the switch really needs to reduce to `ret`.

Adding @GorNishanov to check my understanding.

Reply:
...or @Flakebi

Reply:
`llvm.coro.suspend` is lowered to either 0 in `.resume` function or 1 in .destroy/.cleanup function. 

When `await_suspend` returns a handle, we always want the following call to its `.resume` function to be a tail call in order to have symmetric transfer to work.

Reply:
Right, and moreover the `-1` case should reduce to `ret`, correct?

Reply:
Technically, it depends on the frontend. For C++, yes.

The fancy part here is that symmetric transfer is a C++ feature. We just tried to implement it in the middle end. (This is not good. I know). So we can only consider C++ since the motivation of the patch to save symmetric transfer under pgo.